ksz8863/src/ksz8863_ctrl.c

```diff
@@ -57,56 +59,48 @@ static inline bool bus_unlock(void)
 static esp_err_t ksz8863_i2c_write(uint8_t reg_addr, uint8_t *data, size_t len)
 {
     esp_err_t ret = ESP_OK;
-    i2c_port_t i2c_port = s_ksz8863_ctrl_intf->i2c_bus_spec.i2c_port;
-    uint8_t dev_addr = s_ksz8863_ctrl_intf->i2c_bus_spec.dev_addr;
-
-    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
-    ESP_RETURN_ON_FALSE(cmd, ESP_ERR_NO_MEM, TAG, "I2C link create error");
-    ESP_GOTO_ON_ERROR(i2c_master_start(cmd), err, TAG, "I2C master start error");
-    ESP_GOTO_ON_ERROR(i2c_master_write_byte(cmd, dev_addr | I2C_MASTER_WRITE, true), err, TAG, "I2C master write error");
-    ESP_GOTO_ON_ERROR(i2c_master_write_byte(cmd, reg_addr, true), err, TAG, "I2C master write error");
-    if (likely(reg_addr != KSZ8863_RESET_ADDR)) {
-        ESP_GOTO_ON_ERROR(i2c_master_write(cmd, data, len, true), err, TAG, "I2C master write error");
-    } else {
-        // when SW reset is performed, KSZ does not generate ACK
-        ESP_GOTO_ON_ERROR(i2c_master_write(cmd, data, len, false), err, TAG, "I2C master write error");
-    }
-    ESP_GOTO_ON_ERROR(i2c_master_stop(cmd), err, TAG, "I2C master stop error");
-    // Lock since multiple MAC/PHY instances exist and the KSZ may be also accessed by user (MAC tables, etc...)
-    ESP_GOTO_ON_FALSE(bus_lock(KSZ8863_I2C_LOCK_TIMEOUT_MS), ESP_ERR_TIMEOUT, err, TAG, "I2C bus lock timeout");
-    ESP_GOTO_ON_ERROR(i2c_master_cmd_begin(i2c_port, cmd, pdMS_TO_TICKS(KSZ8863_I2C_TIMEOUT_MS)), err_release, TAG,
-                      "I2C master command begin error");
+    uint8_t *reg_addr_and_data = malloc(len + 1);
+
+    i2c_device_config_t dev_cfg = {
+        .dev_addr_length = I2C_ADDR_BIT_LEN_7,
+        .device_address = s_ksz8863_ctrl_intf->i2c_bus_spec.dev_addr >> 1,
+                .scl_speed_hz = s_ksz8863_ctrl_intf->i2c_bus_spec.scl_speed_hz,
+                .flags.disable_ack_check = (reg_addr == KSZ8863_RESET_ADDR)
+    };
+
+    i2c_master_dev_handle_t dev_handle;
+    ESP_GOTO_ON_ERROR(i2c_master_bus_add_device(s_ksz8863_ctrl_intf->i2c_bus_spec.bus_handle, &dev_cfg, &dev_handle), err, TAG, "Error when trying to add the I2C device");
```

Comment:
This seems to be done only once during init, see https://github.com/espressif/esp-idf/blob/f420609c332fbd2d2f7f188c6579d046c9560e42/examples/peripherals/i2c/i2c_eeprom/components/i2c_eeprom/i2c_eeprom.c#L37. Please investigate.

Reply:
Moved device initialization to `ksz8863_ctrl_intf_init` like the SPI device.